File: module-03-datalink-switching-arp/switch_sim.js
/**
 * Module 03 — JS/Node port of switch_sim.py. Same CAM-table learning,
 * flooding, and VLAN isolation logic, using Buffer instead of struct.
 * Compare this file to switch_sim.py line by line -- the ALGORITHM
 * (learn source, flood unknown/broadcast within VLAN, forward known
 * unicast to one port) is identical; only the byte-layout API differs.
 */

const BROADCAST = Buffer.from("ffffffffffff", "hex");

function macStr(buf) {
  return buf.toString("hex").match(/../g).join(":");
}

function ipToBytes(ip) {
  return Buffer.from(ip.split(".").map(Number));
}

function ipStr(buf) {
  return Array.from(buf).join(".");
}

// --------------------------------------------------------------- frames --
function buildFrame(dstMac, srcMac, ethertype, payload, vlanId = null) {
  if (vlanId !== null) {
    const tag = Buffer.alloc(4);
    tag.writeUInt16BE(0x8100, 0);
    tag.writeUInt16BE(vlanId & 0x0fff, 2);
    const etBuf = Buffer.alloc(2);
    etBuf.writeUInt16BE(ethertype, 0);
    return Buffer.concat([dstMac, srcMac, tag, etBuf, payload]);
  }
  const etBuf = Buffer.alloc(2);
  etBuf.writeUInt16BE(ethertype, 0);
  return Buffer.concat([dstMac, srcMac, etBuf, payload]);
}

function parseFrame(frame) {
  const dst = frame.subarray(0, 6);
  const src = frame.subarray(6, 12);
  let idx = 12;
  let vlanId = null;
  let ethertype = frame.readUInt16BE(idx);
  if (ethertype === 0x8100) {
    const tci = frame.readUInt16BE(idx + 2);
    vlanId = tci & 0x0fff;
    idx += 4;
    ethertype = frame.readUInt16BE(idx);
  }
  idx += 2;
  return { dst, src, ethertype, vlanId, payload: frame.subarray(idx) };
}

// ----------------------------------------------------------------- ARP --
const ETHERTYPE_ARP = 0x0806;
const ARP_REQUEST = 1;
const ARP_REPLY = 2;

function buildArp(op, senderMac, senderIp, targetMac, targetIp) {
  const buf = Buffer.alloc(28);
  buf.writeUInt16BE(1, 0); // htype: Ethernet
  buf.writeUInt16BE(0x0800, 2); // ptype: IPv4
  buf.writeUInt8(6, 4); // hlen
  buf.writeUInt8(4, 5); // plen
  buf.writeUInt16BE(op, 6);
  senderMac.copy(buf, 8);
  ipToBytes(senderIp).copy(buf, 14);
  targetMac.copy(buf, 18);
  ipToBytes(targetIp).copy(buf, 24);
  return buf;
}

function parseArp(payload) {
  return {
    op: payload.readUInt16BE(6),
    senderMac: Buffer.from(payload.subarray(8, 14)),
    senderIp: ipStr(payload.subarray(14, 18)),
    targetMac: Buffer.from(payload.subarray(18, 24)),
    targetIp: ipStr(payload.subarray(24, 28)),
  };
}

// ------------------------------------------------------------- Switch ---
class Switch {
  constructor(portVlan) {
    this.portVlan = portVlan; // Map<port, vlanId>
    this.camTable = new Map(); // key `${vlan}|${mac.hex}` -> port ('|' since MACs contain ':')
    this.hosts = new Map(); // port -> Host
  }

  connect(port, host) {
    this.hosts.set(port, host);
    host.attach(this, port);
  }

  dumpCamTable() {
    console.log("  CAM table:");
    if (this.camTable.size === 0) {
      console.log("    (empty)");
      return;
    }
    for (const [key, port] of this.camTable) {
      const [vlan, mac] = key.split("|");
      console.log(`    VLAN ${vlan.padEnd(3)} ${mac} -> port ${port}`);
    }
  }

  receive(inPort, frameBytes) {
    const parsed = parseFrame(frameBytes);
    const vlan = parsed.vlanId !== null ? parsed.vlanId : this.portVlan.get(inPort);

    const key = `${vlan}|${macStr(parsed.src)}`;
    const learnedBefore = this.camTable.has(key);
    this.camTable.set(key, inPort);
    if (!learnedBefore) {
      console.log(`    [switch] learned ${macStr(parsed.src)} is on port ${inPort} (VLAN ${vlan})`);
    }

    if (parsed.dst.equals(BROADCAST)) {
      console.log(`    [switch] broadcast frame on VLAN ${vlan} -> flooding to VLAN ${vlan} ports only`);
      this._flood(inPort, vlan, frameBytes);
      return;
    }

    const dstKey = `${vlan}|${macStr(parsed.dst)}`;
    if (this.camTable.has(dstKey)) {
      const outPort = this.camTable.get(dstKey);
      console.log(`    [switch] known unicast ${macStr(parsed.dst)} -> forwarding to port ${outPort} ONLY`);
      this._deliver(outPort, frameBytes);
    } else {
      console.log(`    [switch] unknown unicast ${macStr(parsed.dst)} -> flooding VLAN ${vlan} (like broadcast)`);
      this._flood(inPort, vlan, frameBytes);
    }
  }

  _flood(inPort, vlan, frameBytes) {
    for (const [port, portVlan] of this.portVlan) {
      if (port !== inPort && portVlan === vlan) {
        this._deliver(port, frameBytes);
      }
    }
  }

  _deliver(port, frameBytes) {
    const host = this.hosts.get(port);
    if (host) host.receive(frameBytes);
  }
}

// --------------------------------------------------------------- Host ---
class Host {
  constructor(name, macHex, ip) {
    this.name = name;
    this.mac = Buffer.from(macHex.replace(/:/g, ""), "hex");
    this.ip = ip;
    this.arpCache = new Map();
    this.switch = null;
    this.port = null;
  }

  attach(sw, port) {
    this.switch = sw;
    this.port = port;
  }

  send(frameBytes) {
    this.switch.receive(this.port, frameBytes);
  }

  arpRequest(targetIp, vlanId = null) {
    console.log(`  ${this.name}: "who has ${targetIp}? tell ${this.ip}" (broadcast)`);
    const payload = buildArp(ARP_REQUEST, this.mac, this.ip, Buffer.alloc(6), targetIp);
    const frame = buildFrame(BROADCAST, this.mac, ETHERTYPE_ARP, payload, vlanId);
    this.send(frame);
  }

  receive(frameBytes) {
    const parsed = parseFrame(frameBytes);
    if (!parsed.dst.equals(this.mac) && !parsed.dst.equals(BROADCAST)) return;
    if (parsed.ethertype === ETHERTYPE_ARP) this._handleArp(parsed);
  }

  _handleArp(parsed) {
    const arp = parseArp(parsed.payload);
    if (arp.op === ARP_REQUEST && arp.targetIp === this.ip) {
      console.log(`  ${this.name}: that's me -- replying "${this.ip} is at ${macStr(this.mac)}"`);
      const replyPayload = buildArp(ARP_REPLY, this.mac, this.ip, arp.senderMac, arp.senderIp);
      const replyFrame = buildFrame(arp.senderMac, this.mac, ETHERTYPE_ARP, replyPayload, parsed.vlanId);
      this.send(replyFrame);
    } else if (arp.op === ARP_REPLY && !this.arpCache.has(arp.senderIp)) {
      this.arpCache.set(arp.senderIp, arp.senderMac);
      console.log(`  ${this.name}: got ARP reply, caching ${arp.senderIp} -> ${macStr(arp.senderMac)}`);
    }
  }
}

function main() {
  console.log("Building a 4-port switch: ports 1,2 on VLAN 10; ports 3,4 on VLAN 20\n");
  const portVlan = new Map([[1, 10], [2, 10], [3, 20], [4, 20]]);
  const sw = new Switch(portVlan);

  const pcA = new Host("PC-A", "de:ad:be:ef:10:01", "192.0.2.11");
  const pcB = new Host("PC-B", "de:ad:be:ef:10:02", "192.0.2.12");
  const pcC = new Host("PC-C", "de:ad:be:ef:20:01", "192.0.2.21");
  const pcD = new Host("PC-D", "de:ad:be:ef:20:02", "192.0.2.22");

  sw.connect(1, pcA);
  sw.connect(2, pcB);
  sw.connect(3, pcC);
  sw.connect(4, pcD);

  console.log("=".repeat(70));
  console.log("SCENARIO 1: PC-A ARPs for PC-B -- both on VLAN 10");
  console.log("=".repeat(70));
  pcA.arpRequest(pcB.ip);
  console.log();
  sw.dumpCamTable();
  console.log("  PC-A's ARP cache:", Object.fromEntries([...pcA.arpCache].map(([k, v]) => [k, macStr(v)])));

  console.log();
  console.log("=".repeat(70));
  console.log("SCENARIO 2: PC-A ARPs for PC-C's IP -- PC-C is on VLAN 20 (isolated)");
  console.log("=".repeat(70));
  pcA.arpRequest(pcC.ip);
  console.log("  PC-A's ARP cache after:", Object.fromEntries([...pcA.arpCache].map(([k, v]) => [k, macStr(v)])));
  console.log("  Notice: no reply logged above -- same VLAN-isolation result as Python.");

  console.log();
  console.log("=".repeat(70));
  console.log("SCENARIO 3: PC-A sends a second frame to PC-B -- now a KNOWN unicast");
  console.log("=".repeat(70));
  const unicastPayload = Buffer.from("hello PC-B, this is a normal (non-ARP) frame");
  const frame = buildFrame(pcB.mac, pcA.mac, 0x0810, unicastPayload);
  pcA.send(frame);
}

main();

File: module-03-datalink-switching-arp/switch_sim.py
"""
Module 03 -- a real Layer 2 switch's core algorithm, simulated in software:
MAC-address learning (the "CAM table"), flooding of unknown/broadcast
frames, and 802.1Q VLAN tagging that splits one physical switch into
separate broadcast domains.

Every frame here is real bytes (built the same way Module 02 built its
frame, with struct.pack), just delivered by Python function calls instead
of a NIC -- so you can print and inspect every step no hardware could show
you this clearly. All MACs use the same non-routable de:ad:be:ef prefix as
Module 02; IPs use RFC 5737 documentation ranges.
"""
import struct


def mac_str(mac: bytes) -> str:
    return mac.hex(":")


def ip_to_bytes(ip: str) -> bytes:
    return bytes(int(o) for o in ip.split("."))


def ip_str(b: bytes) -> str:
    return ".".join(str(o) for o in b)


BROADCAST = b"\xff" * 6


# --------------------------------------------------------------- frames ---
def build_frame(dst_mac: bytes, src_mac: bytes, ethertype: int, payload: bytes, vlan_id=None) -> bytes:
    if vlan_id is not None:
        tag = struct.pack("!HH", 0x8100, vlan_id & 0x0FFF)  # TPID + VLAN in TCI
        return dst_mac + src_mac + tag + struct.pack("!H", ethertype) + payload
    return dst_mac + src_mac + struct.pack("!H", ethertype) + payload


def parse_frame(frame: bytes) -> dict:
    dst, src = frame[0:6], frame[6:12]
    idx = 12
    vlan_id = None
    ethertype = struct.unpack("!H", frame[idx:idx + 2])[0]
    if ethertype == 0x8100:
        tci = struct.unpack("!H", frame[idx + 2:idx + 4])[0]
        vlan_id = tci & 0x0FFF
        idx += 4
        ethertype = struct.unpack("!H", frame[idx:idx + 2])[0]
    idx += 2
    return {"dst": dst, "src": src, "ethertype": ethertype, "vlan_id": vlan_id, "payload": frame[idx:]}


# ----------------------------------------------------------------- ARP ---
ETHERTYPE_ARP = 0x0806
ARP_REQUEST, ARP_REPLY = 1, 2


def build_arp(op: int, sender_mac: bytes, sender_ip: str, target_mac: bytes, target_ip: str) -> bytes:
    return struct.pack(
        "!HHBBH6s4s6s4s",
        1, 0x0800, 6, 4, op,
        sender_mac, ip_to_bytes(sender_ip),
        target_mac, ip_to_bytes(target_ip),
    )


def parse_arp(payload: bytes) -> dict:
    htype, ptype, hlen, plen, op, sender_mac, sender_ip, target_mac, target_ip = struct.unpack(
        "!HHBBH6s4s6s4s", payload
    )
    return {
        "op": op, "sender_mac": sender_mac, "sender_ip": ip_str(sender_ip),
        "target_mac": target_mac, "target_ip": ip_str(target_ip),
    }


# ------------------------------------------------------------- Switch ----
class Switch:
    def __init__(self, port_vlan: dict):
        self.port_vlan = port_vlan       # port number -> VLAN id
        self.cam_table = {}              # (vlan_id, mac_bytes) -> port number
        self.hosts = {}                  # port number -> Host

    def connect(self, port: int, host: "Host"):
        self.hosts[port] = host
        host.attach(self, port)

    def dump_cam_table(self):
        print("  CAM table:")
        if not self.cam_table:
            print("    (empty)")
        for (vlan, mac), port in self.cam_table.items():
            print(f"    VLAN {vlan:<3} {mac_str(mac)} -> port {port}")

    def receive(self, in_port: int, frame_bytes: bytes):
        parsed = parse_frame(frame_bytes)
        vlan = parsed["vlan_id"] if parsed["vlan_id"] is not None else self.port_vlan[in_port]

        learned_before = (vlan, parsed["src"]) in self.cam_table
        self.cam_table[(vlan, parsed["src"])] = in_port
        if not learned_before:
            print(f"    [switch] learned {mac_str(parsed['src'])} is on port {in_port} (VLAN {vlan})")

        if parsed["dst"] == BROADCAST:
            print(f"    [switch] broadcast frame on VLAN {vlan} -> flooding to VLAN {vlan} ports only")
            self._flood(in_port, vlan, frame_bytes)
            return

        key = (vlan, parsed["dst"])
        if key in self.cam_table:
            out_port = self.cam_table[key]
            print(f"    [switch] known unicast {mac_str(parsed['dst'])} -> forwarding to port {out_port} ONLY")
            self._deliver(out_port, frame_bytes)
        else:
            print(f"    [switch] unknown unicast {mac_str(parsed['dst'])} -> flooding VLAN {vlan} (like broadcast)")
            self._flood(in_port, vlan, frame_bytes)

    def _flood(self, in_port, vlan, frame_bytes):
        for port, port_vlan in self.port_vlan.items():
            if port != in_port and port_vlan == vlan:
                self._deliver(port, frame_bytes)

    def _deliver(self, port, frame_bytes):
        host = self.hosts.get(port)
        if host:
            host.receive(frame_bytes)


# --------------------------------------------------------------- Host ----
class Host:
    def __init__(self, name: str, mac_hex: str, ip: str):
        self.name = name
        self.mac = bytes.fromhex(mac_hex.replace(":", ""))
        self.ip = ip
        self.arp_cache = {}   # ip -> mac
        self.switch = None
        self.port = None

    def attach(self, switch: Switch, port: int):
        self.switch = switch
        self.port = port

    def send(self, frame_bytes: bytes):
        self.switch.receive(self.port, frame_bytes)

    def arp_request(self, target_ip: str, vlan_id=None):
        print(f"  {self.name}: \"who has {target_ip}? tell {self.ip}\" (broadcast)")
        payload = build_arp(ARP_REQUEST, self.mac, self.ip, b"\x00" * 6, target_ip)
        frame = build_frame(BROADCAST, self.mac, ETHERTYPE_ARP, payload, vlan_id)
        self.send(frame)

    def receive(self, frame_bytes: bytes):
        parsed = parse_frame(frame_bytes)
        if parsed["dst"] not in (self.mac, BROADCAST):
            return
        if parsed["ethertype"] == ETHERTYPE_ARP:
            self._handle_arp(parsed)

    def _handle_arp(self, parsed):
        arp = parse_arp(parsed["payload"])
        if arp["op"] == ARP_REQUEST and arp["target_ip"] == self.ip:
            print(f"  {self.name}: that's me -- replying \"{self.ip} is at {mac_str(self.mac)}\"")
            reply_payload = build_arp(ARP_REPLY, self.mac, self.ip, arp["sender_mac"], arp["sender_ip"])
            reply_frame = build_frame(arp["sender_mac"], self.mac, ETHERTYPE_ARP, reply_payload, parsed["vlan_id"])
            self.send(reply_frame)
        elif arp["op"] == ARP_REPLY and arp["sender_ip"] not in self.arp_cache:
            self.arp_cache[arp["sender_ip"]] = arp["sender_mac"]
            print(f"  {self.name}: got ARP reply, caching {arp['sender_ip']} -> {mac_str(arp['sender_mac'])}")


def main():
    print("Building a 4-port switch: ports 1,2 on VLAN 10; ports 3,4 on VLAN 20\n")
    switch = Switch(port_vlan={1: 10, 2: 10, 3: 20, 4: 20})

    pc_a = Host("PC-A", "de:ad:be:ef:10:01", "192.0.2.11")
    pc_b = Host("PC-B", "de:ad:be:ef:10:02", "192.0.2.12")
    pc_c = Host("PC-C", "de:ad:be:ef:20:01", "192.0.2.21")
    pc_d = Host("PC-D", "de:ad:be:ef:20:02", "192.0.2.22")

    switch.connect(1, pc_a)
    switch.connect(2, pc_b)
    switch.connect(3, pc_c)
    switch.connect(4, pc_d)

    print("=" * 70)
    print("SCENARIO 1: PC-A ARPs for PC-B -- both on VLAN 10")
    print("=" * 70)
    pc_a.arp_request(pc_b.ip)
    print()
    switch.dump_cam_table()
    print(f"  PC-A's ARP cache: { {k: mac_str(v) for k, v in pc_a.arp_cache.items()} }")

    print()
    print("=" * 70)
    print("SCENARIO 2: PC-A ARPs for PC-C's IP -- PC-C is on VLAN 20 (isolated)")
    print("=" * 70)
    pc_a.arp_request(pc_c.ip)
    print(f"  PC-A's ARP cache after: { {k: mac_str(v) for k, v in pc_a.arp_cache.items()} }")
    print("  Notice: no reply logged above. The broadcast never reached VLAN 20 at")
    print("  all -- the switch's flood in Scenario 1 & 2 only ever goes to ports")
    print("  whose VLAN matches the frame's VLAN. This IS what VLAN isolation means")
    print("  at the packet level: not a firewall rule, just the switch refusing to")
    print("  flood across VLAN boundaries in the first place.")

    print()
    print("=" * 70)
    print("SCENARIO 3: PC-A sends a second frame to PC-B -- now a KNOWN unicast")
    print("=" * 70)
    unicast_payload = b"hello PC-B, this is a normal (non-ARP) frame"
    frame = build_frame(pc_b.mac, pc_a.mac, 0x0810, unicast_payload)  # arbitrary made-up ethertype
    pc_a.send(frame)
    print("  (compare this log line to Scenario 1's flood -- same source/dest pair,")
    print("   but now the switch forwards to port 2 ONLY, because it's in the CAM table)")


if __name__ == "__main__":
    main()
